Assets/Tile walking mini game/Scripts/Timer.cs

```diff
@@ -0,0 +1,91 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.Events;
+
+public class Timer : MonoBehaviour
+{
+    [SerializeField] private GridManager gridManager;
+
+    [SerializeField] private int timerDuration;
+
+    private int _timeLeft;
+
+    private int timeLeft
+    {
+        get
+        { 
+            return _timeLeft; 
+        }
+
+        set
+        {
+            _timeLeft = value;
+            TimerUpdatedEvent(_timeLeft);
+        }
+    }
+
+    private IEnumerator timerCoroutine;
+
+    public Action<int> TimerUpdatedEvent;
+
+    private bool timerRunning = false;
+
+    private void Awake()
+    {
+        gridManager.GameEventUpdatedEvent += UpdateTimer;
+    }
+
+    private void UpdateTimer(GameState gameState)
+    {
+        if (gameState.CheckCurrentState("new game"))
+            StartTimer();
+
+        else if (gameState.CheckCurrentState("full win") || 
+                 gameState.CheckCurrentState("partial win") || 
+                 gameState.CheckCurrentState("lose"))
+        {
+            StopTimer();
+        }
+    }
+
+    public void StartTimer()
+    {
+        if(timerRunning == true)
+        {
+            Debug.LogError("The timer was already running and you are trying to start it. " +
+                           "This is not allowed.");
+            return;
+        }
+
+        timerCoroutine = RunTimer();
+
+        StartCoroutine(timerCoroutine);
+    }
+
+    public void StopTimer()
+    {
+        if(timerCoroutine != null)
```

Comment:
I guess you could just check if `timerRunning` is `true` here rather than null checking the `timerCoroutine`

I wasn't sure if we needed the `timerCoroutine` field